File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/firestore/StateHelper.java
package com.klnvch.greenhousecommon.firestore;

import androidx.annotation.NonNull;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Map;

public abstract class StateHelper {
    public static final String FIELD_DEVICE_ID = "deviceId";
    public static final String FIELD_TIME = "time";
    private final ObjectMapper mapper;

    public StateHelper() {
        mapper = new ObjectMapper();
        //mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    @NonNull
    public Map<String, Object> toMap(@NonNull Object object) {
        return mapper.convertValue(object, new TypeReference<Map<String, Object>>() {
        });
    }

    @NonNull
    public  <T> T fromMap(@NonNull Map<String, Object> data, Class<T> classOfT) {
        return mapper.convertValue(data, classOfT);
    }

    public abstract String getCollectionPath();
}

File: GreenHouseController/greenhousecommon/src/main/java/com/klnvch/greenhousecommon/firestore/StateHelper.kt
package com.klnvch.greenhousecommon.firestore

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper

abstract class StateHelper {
    private val mapper: ObjectMapper = ObjectMapper()

    fun toMap(data: Any): Map<String, Any> {
        return mapper.convertValue(data, object : TypeReference<Map<String, Any>>() {})
    }

    fun <T> fromMap(data: Map<String, Any>, classOfT: Class<T>): T {
        return mapper.convertValue(data, classOfT)
    }

    abstract val collectionPath: String

    companion object {
        const val FIELD_DEVICE_ID = "deviceId"
        const val FIELD_TIME = "time"
    }

    init {
        //mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }
}
